### PR TITLE
Tensor Cores 2: Local Buffers Edition

### PR DESCRIPTION
**Pull request by geohot**

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/codegen/cstyle.py                 3               2
tinygrad/codegen/linearizer.py            59               4
------------------------------------------------------------
total                                     62               6
------------------------------------------------------------
lines added in the tinygrad folder: 56
```

File: tinygrad/codegen/linearizer.py [hunk 1]
```diff
@@ -3,22 +3,24 @@
...
-class UOps(Enum): LOOP = auto(); DEFINE_LOCAL = auto(); LOAD = auto(); ALU = auto(); CONST = auto(); ENDLOOP = auto(); STORE = auto(); CAST = auto(); BARRIER = auto(); \
+# bottom ones are asm only
+class UOps(Enum): LOOP = auto(); DEFINE_LOCAL = auto(); LOAD = auto(); ALU = auto(); CONST = auto(); ENDLOOP = auto(); STORE = auto(); CAST = auto(); BARRIER = auto(); WMMA = auto(); \
                   SPECIAL = auto(); DEFINE_REGISTER = auto(); LABEL = auto(); COND_BRANCH = auto() # noqa: E702
 
 class LocalBuffer(NamedTuple):
   name: str
   size: int
   dtype: DType = dtypes.float32
+  alias: Optional[LazyBuffer] = None
```
Comment (geohot) [inline tinygrad/codegen/linearizer.py:23]: Remove this?

File: tinygrad/codegen/cstyle.py [hunk 2]
```diff
@@ -103,6 +114,14 @@ def kk(s): kernel.append("  "*depth+s)
           pend_close = None
         depth -= 1
         kk("}"*len(args[0]) + f" /* {args[1]} */")
+    elif uop == UOps.WMMA:
```
Comment (geohot) [inline tinygrad/codegen/cstyle.py:117]: Refactor the input of these to be _float2?

**Comment by atyshka** [comment]
Would you mind sharing the logic behind the reshape gymnastics used here? I'm interested in integrating NVIDIA support, but I'm trying to make sense of why there's a need to add all the intermediate dimensions via upcasting, as long as the local dimensions are right for the tensor cores (in this case, 64 threads)

**Pull request merged**